### PR TITLE
dispatch an event the user can handle to add additional metadata

### PR DESCRIPTION
**Pull request by chris48s**
See discussion on issue #6 
btw, I noticed while I was doing this that the documentation link on line 7 of your CONTRIBUTING.md is a dead link.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +89,49 @@ You can optionally retrieve all the versions:
 $versions = $entity->versions();
 ```
 
+### Storing Additional MetaData
+
+`cakephp-version` dispatches an event `Model.Version.beforeSave` which you can optionally handle to attach additional meta-data about the version.
```
Comment (josegonzalez) [inline README.md]: Can you show how someone might be able attach the listener (just for clarity's sake).

File: README.md [hunk 2]
```diff
@@ -89,6 +89,49 @@ You can optionally retrieve all the versions:
...
+CREATE TABLE `version` (
+    `id` int(11) NOT NULL AUTO_INCREMENT,
+    `version_id` int(11) DEFAULT NULL,
+    `model` varchar(255) NOT NULL,
+    `foreign_key` int(10) NOT NULL,
+    `field` varchar(255) NOT NULL,
+    `content` text,
+    `created` datetime NOT NULL,
+    `custom_field1` varchar(255) NOT NULL,
+    `custom_field2` varchar(255) NOT NULL,
```
Comment (josegonzalez) [inline README.md]: Use an sql comment to show that these two fields are custom.

File: src/Model/Behavior/VersionBehavior.php [hunk 3]
```diff
@@ -161,6 +162,13 @@ public function beforeSave(Event $event, Entity $entity, ArrayObject $options)
                 'content' => $content,
                 'created' => $created,
             ];
+
+            $event = new Event('Model.Version.beforeSave', $this, $options);
+            $user_data = EventManager::instance()->dispatch($event);
```
Comment (josegonzalez) [inline src/Model/Behavior/VersionBehavior.php]: `$userData`

File: src/Model/Behavior/VersionBehavior.php [hunk 4]
```diff
@@ -161,6 +162,13 @@ public function beforeSave(Event $event, Entity $entity, ArrayObject $options)
...
+            if (isset($user_data->result) && is_array($user_data->result)) {
+                $data = array_merge($data, $user_data->result);
```
Comment (josegonzalez) [inline src/Model/Behavior/VersionBehavior.php]: We should note that the custom data will be merged with the versioned data, and as such people should be careful about what they merge in.

**Comment by chris48s** [comment]
Wow - that's some speedy feedback. Also it looks like I have some test failures under Postgres which I had not considered. Let me take a look at these various issues and get back to you with an updated PR.

**Comment by chris48s** [comment]
Hopefully this is ok now. If there's anything else you need I will pick it up after Christmas.
Thanks

**Comment by josegonzalez** [comment]
Merged and released!

**Pull request merged**